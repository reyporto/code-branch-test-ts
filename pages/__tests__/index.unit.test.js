import TestRenderer from 'react-test-renderer'
import { cleanup, render, fireEvent, act } from '@testing-library/react'
import Home from '..'

const mockResponse = {
    balance: true,
}

global.fetch = jest.fn(() =>
    Promise.resolve({
        json: () => Promise.resolve(mockResponse),
    })
)

describe('Home Component', () => {
    afterEach(cleanup)

    beforeEach(() => {
        jest.clearAllMocks()
    })

    test('should render Home component', () => {
        const component = TestRenderer.create(<Home />)
        expect(component.toJSON()).toMatchSnapshot()
    })

    test('should submit success - is balanced', async () => {
        const { getByRole, getByPlaceholderText } = render(<Home />)

        const codeInput = getByPlaceholderText(/Code/i)
        act(() => {
            fireEvent.change(codeInput, { target: { value: 'const addOne = value => { return value + 1 }' } })
        })

        const submit = getByRole('button', { name: 'Submit' })
        await act(async () => {
            fireEvent.submit(submit)
        })
    })

    test('should submit success - is not balanced', async () => {
        const { getByRole, getByPlaceholderText } = render(<Home />)
        mockResponse.balance = false

        const codeInput = getByPlaceholderText(/Code/i)
        act(() => {
            fireEvent.change(codeInput, { target: { value: 'const addOne = value => { return value + 1 }}' } })
        })

        const submit = getByRole('button', { name: 'Submit' })
        await act(async () => {
            fireEvent.submit(submit)
        })
    })

    test('should submit exception', async () => {
        const { getByRole, getByPlaceholderText } = render(<Home />)

        global.fetch = jest.fn(() =>
            Promise.reject(new Error('Error!'))
        )

        const codeInput = getByPlaceholderText(/Code/i)
        act(() => {
            fireEvent.change(codeInput, { target: { value: 'const addOne = value => { return value + 1 }}' } })
        })

        const submit = getByRole('button', { name: 'Submit' })
        await act(async () => {
            fireEvent.submit(submit)
        })
    })
})