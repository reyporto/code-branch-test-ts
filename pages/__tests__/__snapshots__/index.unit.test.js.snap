// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component should render Home component 1`] = `
<div
  className="container"
>
  <main
    className="main"
  >
    <h1
      className="title"
    >
      Welcome to Brackets Balance
    </h1>
    <p
      className="description"
    >
      To get started write
       
      <code
        className="code"
      >
        code
      </code>
    </p>
    <div
      className="grid"
    >
      <form
        onSubmit={[Function]}
      >
        <textarea
          className="textarea"
          id="code"
          name="code"
          onChange={[Function]}
          placeholder="Code"
          value=""
        />
        <br />
        <button
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </main>
  <footer
    className="footer"
  >
    <p>
      @Reinaldo Porto
    </p>
  </footer>
</div>
`;
